### PR TITLE
Move all bash custom completions to Go

### LINKED ISSUE #882: Implement kubectl custom command-line completions using go instead of bash
**What would you like to be added**:
* Implement dynamic completions (i.e. show list of pods) using go instead of bash using new capabilities available in Cobra 1.0.0.
* Remove all existing custom command-line completion bash and zsh workaround code and let cobra generate the completion script.
* Add unit tests for each of the completion types (node, context, cluster, resource, etc)
* There should be no impact to how completions are used or how they are installed, so this change should be transparent to the end-user.

**Why is this needed**:
The Cobra reference was recently upgraded to version 1.0.0 for the Kubernetes 1.19 release. This version of Cobra has a [new, recommended way to specify dynamic completions using Go](https://github.com/spf13/cobra/blob/master/bash_completions.md#dynamic-completion-of-nouns).

> Note: Custom Completions written in Go will automatically work for other shell-completion scripts (e.g., Fish shell), while Custom Completions written in Bash will only work for Bash shell-completion. It is therefore recommended to use Custom Completions written in Go.

By implementing completions in Go (instead of bash), we can get rid of the [custom bash code](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/cmd.go#L82-L289) to handle completions which doesn't have tests, and we can also get rid of the [zsh workaround code](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/completion/completion.go#L146-L269) because Cobra 1.0.0 supports zsh completions now.

Making this change will also enable kubectl to support fish completion, [which has been requested](https://github.com/kubernetes/kubectl/issues/576), as Cobra 1.0.0 supports fish completions.

Writing completions in go could also make it easier to optimize completions to handle large resource lists which [have been found to be slow in some cases](https://github.com/kubernetes/kubectl/issues/683) because we will have more control over how the requests are being performed internally.

### PR DESCRIPTION
**Pull request by marckhouzam**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Based on #93714, this PR finishes removing all bash completion scripting and replaces it fully with Go completions.

*Advantages:*
* easier maintenance of custom completions
* ability to write Go tests for custom completions
* allow to eventually move to native zsh completion
* allow Fish shell completion PR (#92989) to fully support all of kubectl's custom completions
* removes lack of portability of bash scripts
* will allow to improve existing custom completions and add other ones

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubectl/issues/882

**Special notes for your reviewer**:

The PR is based on top of @knight42 great work of #93714.  However, it has been mostly refactored when included with the rest of the PR.

The pattern that the PR follows is:
- main completion functions are added in their matching files:
    - completion of api resource names: `apiresources.go#CompGetResourceList(...)`
    - completion of individual resources: `get.go#CompGetResource(...)`
- a new `completion.go` file is added to the `util` package with the new completion code
- the Cobra `ValidArgsFunction` field is added to every cobra command that was part of the bash `__kubectl_custom_func`.  Each `ValidArgsFunction` calls the `util` package.
- flag completion for `--namespace`, `--context`, `--user`, `--cluster` which are global was added in `cmd.go`
- flag completion for `--clusterrole` which is limited to the `create clusterrolebinding` command was added in `create_clusterrolebinding.go`
- go tests are added in the `util` package in `completion_tests.go`

**Test the new completion code**
* build kubectl from source
```
make WHAT=cmd/kubectl
```
* generate completion code
```
# zsh
source <(./_output/bin/kubectl completion zsh)

# bash
source <(./_output/bin/kubectl completion bash)
```
* use it
*Please note that for the new completion code to work, one must use the new kubectl on the command-line.*
```
./_output/bin/kubectl get [TAB]
# debug
./_output/bin/kubectl __complete get ""<ENTER>

./_output/bin/kubectl describe -n kube-system pod [TAB]
```

**Does this PR introduce a user-facing change?**:

Regression:

Custom completions don't work following concatenated short flags, e.g., `kubectl exec -it <TAB>`.
This will be fixed in the next Cobra release: https://github.com/spf13/cobra/pull/1258

Only minor changes:
- File completion is now disabled for the Go completions as it does not make sense for these cases
- Completion for the `cp` command works slightly differently; instead of doing file completion all the time, which I found to be very verbose, I have used the same solution that `helm` implemented, which is to only show `/` and `./` to indicate file completion is valid, and then to trigger full file completion once the user starts typing something.  You can try this by comparing: `./_output/bin/kubectl cp <TAB>` and `./_output/bin/kubectl cp /tmp/<TAB>`
- Invalid command-lines will not lead to completions, for example:
  `kubectl autoscale deployment --max NaN <TAB>`
will trigger file completion (that is what Cobra does when there is an error in the command-line) instead of accepting the invalid flag value and continuing with completion.

```release-note
Shell completion has been migrated to Cobra's go solution.  `kubectl` is now smarter about disabling file completion when it does not apply.  Furthermore, completion for the `cp` command does not show all files unless the user has started typing something.
```

### CODE REVIEW

**Comment by marckhouzam** [comment]
I've added instructions on how to test, taken from #93714

You may be interested @brianpursley @eddiezane @knight42 @WLun001

**Comment by marckhouzam** [comment]
This is finally ready for review.  Please see the updated descriptions for details.

Although the PR seems very large, there is a lot of repetitive code following a simple pattern.

I wasn't able to implement all go tests I wanted to have.  But if someone more experienced can help, it shouldn't be very hard (help needed described in the PR description).

**Comment by marckhouzam** [comment]
This PR is quite large.
Would it be helpful if I split it into independent PRs to make it easier to review?

**Comment by marckhouzam** [comment]
Support for comma-separated lists for `kubectl get` has just been added by #98301.  This PR needs to be updated to include this as well.  I'll try to get to it.

**Comment by eddiezane** [comment]
@marckhouzam this will need a rebase to remove bazel. That will help cut down on the review size.

**Comment by marckhouzam** [comment]
This is ready with Bazel removed.  Thanks @eddiezane  for pointing it out.

There are 3 commits:
1. The original commit from #93714 from @knight42, since that is what I started from
2. The conversion of all bash code into Go code
3. The additions of Go tests for the new Go code

The conversion from bash to Go mostly follows a simple pattern of either listing resource types or listing resource names.
Please see the description of the PR for details on this pattern.

Some tests are commented out as I didn't have the knowledge to make them all work.
There are a couple of `//TODO` and `//FIXME` where help is needed.

**Comment by marckhouzam** [comment]
Oh, and the rebase includes the recent feature of #98301 (which didn't work for zsh before, but in Go code, it does)

**Comment by rikatz** [comment]
@eddiezane sorry, saw this on meeting notes, will put on my queue to proper review and test!

**Comment by rikatz** [comment]
Hey @marckhouzam thanks for the PR.

I can't say I am pretty aware of how cobra completions works, so I'll put some reviews from my user perspective points here:

I'm testing everything on zsh here

* So far, kubectl tab works fine, returning all the subcommands, which is great
* It seems that some shortcuts are not working/being rendered. The first that I tried was "-n" (namespace) in a new installed zsh. Maybe missing in "cmd.go"?
* I've made an attempt with bash as well (5.0) but - _output/bin/kubectl get po_get_comp_words_by_ref: command not found

This is my initial test. Will make some further tests during the week

**Comment by hjpotter92** [comment]
I've been following this issue for quite sometime now. The reason I look forward to a different mechanism for autocompletion is because when I have the `kubectl` command wrapped with a 3rd party tool for formatting purposes, it is unable to fetch complete list of completion candidates for the sub-commands on zsh.

The wrapper is [grc](https://github.com/garabik/grc) as follows:

```shell
kubectl () {
  grc --colour=auto ${commands[$0]} "$@"
}
```

Now, the tab completion does not fetch candidates for a command such as `kubectl get <TAB>`, as shown in the following screencast (my omz setup autoloads the wrapper utlity)

[asciicast](https://asciinema.org/a/0fv9piuBYDNthXHsoRiblSBVb)

**Comment by marckhouzam** [comment]
> Hey @marckhouzam thanks for the PR.

Thanks @rikatz  for trying this out.

> I'm testing everything on zsh here
> 
> * So far, kubectl tab works fine, returning all the subcommands, which is great

For subcommands and flags nothing has changed from the current behavior.

> * It seems that some shortcuts are not working/being rendered. The first that I tried was "-n" (namespace) in a new installed zsh. Maybe missing in "cmd.go"?
> * I've made an attempt with bash as well (5.0) but - _output/bin/kubectl get po_get_comp_words_by_ref: command not found

Do you have the bash_completion package installed?  It could explain both problems maybe.

**Comment by marckhouzam** [comment]
@hjpotter92 Out of curiosity, I've tried to reproduce your problem.  I haven't installed grc but I tried to do something similar and for me changing your method very slightly makes the completion work.  Do you want to try using the following to see if it works better:
```
kubectl () {
  grc --colour=auto kubectl "$@"
}
```

**Comment by rothgar** [comment]
I've also been following this PR for a while as the use of bash_completion is breaking some of my zsh completion customizations.

I just wanted to mention if you want color output you may want to check out [`kubecolor`](https://github.com/dty1er/kubecolor) which is a wrapper with specific colorization for kubectl output.

**Comment by marckhouzam** [comment]
> > * It seems that some shortcuts are not working/being rendered. The first that I tried was "-n" (namespace) in a new installed zsh. Maybe missing in "cmd.go"?
> > * I've made an attempt with bash as well (5.0) but - _output/bin/kubectl get po_get_comp_words_by_ref: command not found
> 
> Do you have the bash_completion package installed? It could explain both problems maybe.

@rikatz to clarify this, this PR retains the current zsh completion which uses the bash script.  That is why you still need to have the bash_completion package installed, even for zsh.  If this PR gets merged, we will then be able to use native zsh completion which will no longer need the bash_completion package.

A good way to compare this PR with the current behaviour would be for you to have two running shells, one using the current kubectl and its completion, and the other using the new kubectl using this PR and its completion.  You can then verify that things behave the same (except all written in Go :-)).

File: staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go
Comment (phil9909) [inline staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go:149]: Isn't that the default behavior? Why is this re-implemented here?

**Review by phil9909 — COMMENTED** [review]
I noticed there is quite some duplication in the implementation of `ValidArgsFunction`.
A lot of these are a bit different, but many of them seem to complete for `<resource kind> <resource name>`. Maybe this should be refactored.

I usually apply "DRY" for duplication of `n > 3` even for small pieces of code.
It always comes with the risk of "drifting apart", where a feature is only added in some copies of the code.

**Comment by brianpursley** [comment]
I haven't gotten time to review this until now and I apologize for that.  I think the volume of files touched makes it a little unwieldy to review.

I've gone through about half of the files so far and in general it looks like a good improvement.  I tested some of the commands locally using bash and zsh and completions seem to work.  The old bash completion script is gone, so that would be a win for maintainability going forward.

I do think we need to decide whether the FIXMEs and TODOs are something that would prevent this from being merged or if they need to be fixed before merging.  I would draw the line at any loss of functionality myself, as @phil9909 mentioned in a comment above, we have no tests for completions currently, so not having 100% coverage of completions probably would not block this from being merged.

All that being said, **I do have a suggestion that I think would greatly reduce the footprint of this PR:**

What if most, if not all, of the completions are refactored into a single `completions.go` file under the util folder or similar common location?

The reason I ask this is there is a lot of repetition where many commands have the exact same completion.  For example, the following pattern of completing the first argument as resource types, and second argument as resource names is used in 8 different places:

```
ValidArgsFunction: func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
			var comps []string
			if len(args) == 0 {
				comps = apiresources.CompGetResourceList(f, cmd, toComplete)
			} else if len(args) == 1 {
				comps = get.CompGetResource(f, cmd, args[0], toComplete)
			}
			return comps, cobra.ShellCompDirectiveNoFileComp
		},
```

Instead, you could create a single function that returns a completion function for a given factory into `util/completions.go`, something like this:

```
func ResourceTypeAndNameCompletionFunction(f cmdutil.Factory) func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	return func (cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
		var comps []string
		if len(args) == 0 {
			comps = apiresources.CompGetResourceList(f, cmd, toComplete)
		} else if len(args) == 1 {
			comps = get.CompGetResource(f, cmd, args[0], toComplete)
		}
		return comps, cobra.ShellCompDirectiveNoFileComp
	}
}
```

Maybe there is a more elegant way to do this in go without having a function that returns a function, but I think you can probably see what I'm getting at anyway.  

With this function, in the 8 different places it is needed, it can just do this instead:
```
ValidArgsFunction: util.ResourceTypeAndNameCompletionFunction(f),
```

Besides eliminating duplicate code, this also has the benefit of only needing to be tested in `completions_test.go` and not in each individual command.

I'm thinking this would greatly reduce this size of this PR and help focus the attention on the actual completion functions instead of on each individual command's implementation of completions.

Of course there still might be some highly custom scenario (ie. `cp`) where it needs to have its own completion that is unique to its purpose, but I think for 90% of the commands the completions are going to be the same as some other command's completions.

File: staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go
Comment (brianpursley) [reply staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go:114]: Does current kubectl cp complete remote files?

File: staging/src/k8s.io/kubectl/pkg/cmd/config/config.go
Comment (marckhouzam) [reply staging/src/k8s.io/kubectl/pkg/cmd/config/config.go:154]: Excellent point!  Will merge your suggestion.

File: staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go
Comment (marckhouzam) [reply staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go:114]: Nope, `kubectl cp` does not currently complete remove files.  The `TODO` is actually copied from the current bash code that is being replaced.

**Comment by marckhouzam** [comment]
> I haven't gotten time to review this until now and I apologize for that. I think the volume of files touched makes it a little unwieldy to review.

Yes, the size of the PR definitely makes is harder to review.

> I do think we need to decide whether the FIXMEs and TODOs are something that would prevent this from being merged or if they need to be fixed before merging. I would draw the line at any loss of functionality myself, as @phil9909 mentioned in a comment above, we have no tests for completions currently, so not having 100% coverage of completions probably would not block this from being merged.

That is fine with me.  The FIXME and TODOs that are left, as you point out, are not critical and could always be improved in a future PR.

> All that being said, **I do have a suggestion that I think would greatly reduce the footprint of this PR:**
> 
> What if most, if not all, of the completions are refactored into a single `completions.go` file under the util folder or similar common location?
> [...]

This makes sense to me.  The repeated patterns grew as I was progressing through the PR and replacing bash code.
I will consolidate as much as possible as @brianpursley and @phil9909 have suggested.

**Comment by marckhouzam** [comment]
Be aware that there is a regression with this change.  Custom completions won't work following concatenated short flags.  For example `kubectl exec -it <TAB>` (while `kubectl exec -i -t <TAB>` will work).

This is a bug in Cobra for which the fix was merged today: https://github.com/spf13/cobra/pull/1258
I expect a new Cobra release in May that will address this bug and other small completion bugs.

**Comment by marckhouzam** [comment]
@brianpursley I have tried to follow your suggestion in a new commit.  There is now a new `pkg/util/completion.go` file which contains the repeated completion code.  It has allowed to reduce about 200 new lines, and made the code clearer I believe.

I was not able to apply this pattern to the `use_context`, `delete_context`, `rename_context` and `delete_cluster` commands, as they don't have the factory available to pass to the `util` package, and because their completion code is partially used in some flag completion as well.  I also didn't apply the pattern to the `cp` command which has its own, more complex completion code.  Finally, and most interestingly, I couldn't apply the pattern to the `get` command. The `get` command would have to call the `util` package, but the `util` package needs to call the `get` package to be able to list the resource names for the completion; this created a circular dependancy, so I left the `get` command with its own `ValidArgsFunction`.  If you have a suggestion to improve that, please let me know. 

Finally, I didn't change the tests. I am not sure I understood what you meant when you suggested testing this code in a single `completion_test.go` file.  I believe we need to test that each command, even if it re-uses code, does properly perform completion.  So, although the `describe` and the `annotate` commands call the same new `util.ResourceTypeAndNameCompletionFunc(f)`, I feel we should still test both commands independently to make sure each command's `ValidArgsFunction` is not changed erroneously.  But maybe I misunderstood your suggestion?

Do let me know if this is in-line with what you had in mind and if I can improve things more.

Thanks

**Comment by brianpursley** [comment]
> @brianpursley I have tried to follow your suggestion in a new commit. There is now a new `pkg/util/completion.go` file which contains the repeated completion code. It has allowed to reduce about 200 new lines, and made the code clearer I believe.

Looking better... it is getting a little lest repetitive, which is good.

See my comments below regarding some of your questions.

> 
> I was not able to apply this pattern to the `use_context`, `delete_context`, `rename_context` and `delete_cluster` commands, as they don't have the factory available to pass to the `util` package, and because their completion code is partially used in some flag completion as well. 

These should be even simpler to move into `completion.go` because they don't have a factory.  Can you just do this in `completion.go`?
```
func ContextCompletionFunc(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
	if len(args) == 0 {
		return config.CompListContextsInConfig(nil, toComplete), cobra.ShellCompDirectiveNoFileComp
	}
	return nil, cobra.ShellCompDirectiveNoFileComp
}
```

Then in `use_context.go` it just becomes this line:
```
ValidArgsFunction: util.ContextCompletionFunc,
```

Similar thing for `delete_context`, `rename_context`, and `delete_cluster`.

> I also didn't apply the pattern to the `cp` command which has its own, more complex completion code. 

`cp` is sort of its own beast completion-wise, so it probably makes sense to stay in `cp.go` like you have it. It is so unique to the `cp` command that reusability isn't really a factor here.

> Finally, and most interestingly, I couldn't apply the pattern to the `get` command. The `get` command would have to call the `util` package, but the `util` package needs to call the `get` package to be able to list the resource names for the completion; this created a circular dependancy, so I left the `get` command with its own `ValidArgsFunction`. If you have a suggestion to improve that, please let me know.

Yeah, I see what you're saying. Ideally, I think `CompGetResource()`, `CompGetContainers()`, and `CompGetFromTemplate()` would exist and be encapsulated within `completion.go`.  But I see you are leveraging the existing get functionality to perform the calls to get completion data.  There's probably some extra stuff in there that is not needed for completions though.  I wonder if there could be a completion-specific versions `CompGetResource()` and `CompGetContainers()` that do not rely on the code in `get.go` but are entirely self-contained.

I don't know, maybe that is beyond the scope of this PR.  It is probably OK to go with how it is now and maybe follow-up with a refactor later to encapsulate the completion data fetching logic into `completion.go` so that it can avoid a circular reference and also to potentially streamline/optimize it.  Thoughts on this from other reviewers?
 
> Finally, I didn't change the tests. I am not sure I understood what you meant when you suggested testing this code in a single `completion_test.go` file. I believe we need to test that each command, even if it re-uses code, does properly perform completion. So, although the `describe` and the `annotate` commands call the same new `util.ResourceTypeAndNameCompletionFunc(f)`, I feel we should still test both commands independently to make sure each command's `ValidArgsFunction` is not changed erroneously. But maybe I misunderstood your suggestion?

My point was that if we test `ResourceTypeAndNameCompletionFunc(f)` in completion_test.go one time, and we know it works in the general sense, then we should not need to test it over and over for each individual command that uses it.

Let's take `annotate.go` for example.  The only change you made is adding this one line:

```
ValidArgsFunction:     util.ResourceTypeAndNameCompletionFunc(f),
```

I think we can assume, for the purposes of unit testing, that Cobra itself will work properly. We should not need to test that Cobra is able to provide argument completion, that is for Cobra to test.  It is enough for us to only test Kubernetes-specific code to make sure it works as expected.

I don't think it is necessarily wrong to test in every command, and maybe others will disagree with me (please feel free to comment if you do), but I think it is not needed because it is mostly just testing Cobra at that point.  There isn't any testable logic in that one line of code you are adding.

**Comment by marckhouzam** [comment]
> These should be even simpler to move into `completion.go` because they don't have a factory. Can you just do this in `completion.go`?
> 
> ```
> func ContextCompletionFunc(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
> 	if len(args) == 0 {
> 		return config.CompListContextsInConfig(nil, toComplete), cobra.ShellCompDirectiveNoFileComp
> 	}
> 	return nil, cobra.ShellCompDirectiveNoFileComp
> }
> ```

Of course!  Not sure why I didn't think of it.

Ah, but looking in more detail, this will cause another circular dependency where `config` will call `util.ContextCompletionFunc` but `util` will call `config.CompListContextsInConfig`.  Ok, so I will just move `CompListContextsInConfig()` into the `util` package as well to avoid the circular dependency.
Oh, but now I'm missing the factory 😢 .
I'll add a `util.SetFactoryForCompletion(f cmdutil.Factory)` function to store the factory globally in the `util` package.
I don't like this very much, so suggestions would be appreciated, if any.

> `cp` is sort of its own beast completion-wise, so it probably makes sense to stay in `cp.go` like you have it. It is so unique to the `cp` command that reusability isn't really a factor here.

👍 

> > Finally, and most interestingly, I couldn't apply the pattern to the get command. The get command would have to call the util package, but the util package needs to call the get package to be able to list the resource names for the completion; this created a circular dependancy, so I left the get command with its own ValidArgsFunction. If you have a suggestion to improve that, please let me know.

> Yeah, I see what you're saying. Ideally, I think CompGetResource(), CompGetContainers(), and CompGetFromTemplate() would exist and be encapsulated within completion.go. But I see you are leveraging the existing get functionality to perform the calls to get completion data. There's probably some extra stuff in there that is not needed for completions though. I wonder if there could be a completion-specific versions CompGetResource() and CompGetContainers() that do not rely on the code in get.go but are entirely self-contained.

> I don't know, maybe that is beyond the scope of this PR. It is probably OK to go with how it is now and maybe follow-up with a refactor later to encapsulate the completion data fetching logic into `completion.go` so that it can avoid a circular reference and also to potentially streamline/optimize it. Thoughts on this from other reviewers?

I'll leave as is and see if other opinions come up.

> > Finally, I didn't change the tests.

> Let's take `annotate.go` for example. The only change you made is adding this one line:
> 
> ```
> ValidArgsFunction:     util.ResourceTypeAndNameCompletionFunc(f),
> ```
> I don't think it is necessarily wrong to test in every command, and maybe others will disagree with me (please feel free to comment if you do), but I think it is not needed because it is mostly just testing Cobra at that point. There isn't any testable logic in that one line of code you are adding.

My concern was that if the `ValidArgsFunction` line was removed from `annotate.go` by mistake, say in a refactor, then there would be no test failure, although completion would be broken for `annotate`.  Same idea if the wrong `util` function is being called by `ValidArgsFunction`.

But maybe I'm being overly cautious?

**Comment by marckhouzam** [comment]
@brianpursley The last forced-push following the proposed refactoring.  If we ignore the tests, the code should be quite simple to review.

As for the tests,  I'll wait for confirmation if testing individual commands for completion is  or is not required, before cleaning up.

**Comment by marckhouzam** [comment]
> As for the tests, I'll wait for confirmation if testing individual commands for completion is or is not required, before cleaning up.

The tests should not keep delaying this PR, so I reverted the original commit which added the tests, and I wrote a few tests in the `util` package in `completion_tests.go` as suggested by @brianpursley.  The original tests are still available in the commit history for reference.

Also,  as a way to maybe to stir up some momentum, I moved to native zsh completion and removed all that extra hack-y shell code (see https://github.com/kubernetes/kubernetes/pull/96087/commits/77bb053102cbe1581857dd35e9c095888d2c85d3).
With this, zsh gets completion descriptions:
```
$ source /tmp/kubectl completion zsh
$ /tmp/kubectl p<TAB>
patch         -- Update field(s) of a resource
plugin        -- Provides utilities for interacting with plugins.
port-forward  -- Forward one or more local ports to a pod
proxy         -- Run a proxy to the Kubernetes API server
```

Hopefully, we can get this merged.
/cc @brianpursley @rikatz @eddiezane

**Review by soltysh — APPROVED** [review]
A few nits that can be addressed in a followup. Apologies for the delays in reviewing and thank you very much for this great effort to you @marckhouzam and @brianpursley and @phil9909 for reviews.
/lgtm
/approve

**Comment by sttts** [comment]
😍

**Comment by marckhouzam** [comment]
Thanks @soltysh!  I will soon address your comment.

Should I revert the change that moves zsh completion to being native from this PR and leave it for a different PR?
I'm asking because I believe the zsh change will require an update to the documentation; specifically, aliases will be handled automatically by zsh, unlike what is mentioned here: https://kubernetes.io/docs/tasks/tools/included/optional-kubectl-configs-zsh/

**Comment by marckhouzam** [comment]
Oh and there is still a regression with this change caused by https://github.com/spf13/cobra/pull/1258

I'm still waiting for the next Cobra release: https://github.com/spf13/cobra/issues/1388

**Pull request merged**
